src/test/groovy/com/gooddata/executeafm/afm/PopMeasureDefinitionTest.groovy

```diff
@@ -17,64 +19,105 @@ import static spock.util.matcher.HamcrestSupport.that
         when:
-        def measure = new PopMeasureDefinition("mId", new IdentifierObjQualifier("id"), inputOffset)
+        def measure = new PopMeasureDefinition("mid", new IdentifierObjQualifier("id"))
         def copy = measure.withObjUriQualifier(new UriObjQualifier("uri"))
 
         then:
-        copy.measureIdentifier == 'mId'
         copy.objQualifier.uri == 'uri'
-        copy.offset == expectedOffset
+    }
 
-        where:
-        inputOffset | expectedOffset
-        null        | null
-        -2          | -2
+    def "should copy with uri converter"() {
+        when:
+        def measure = new PopMeasureDefinition("mid", new IdentifierObjQualifier("id"))
+        def qualifiersConversionMap = [(new IdentifierObjQualifier("id")): new UriObjQualifier("uri")]
+        def copy = measure.withObjUriQualifiers({ identifierQualifier ->
+            return Optional.ofNullable(qualifiersConversionMap.get(identifierQualifier))
+        })
+
+        then:
+        copy.popAttribute.uri == 'uri'
     }
 
-    def "test serializable"() {
+    def "should return the same object when copying with uri converter and uri is already used"() {
         when:
-        PopMeasureDefinition measureDefinition = readObjectFromResource("/$jsonMeasureDefinition", PopMeasureDefinition)
-        PopMeasureDefinition deserialized = SerializationUtils.roundtrip(measureDefinition)
+        def definition = new PopMeasureDefinition("mid", new UriObjQualifier("uri"))
+        def copy = definition.withObjUriQualifiers({ identifierQualifier -> Optional.empty() })
+
+        then:
+        definition == copy
+    }
+
+    def "should fail when qualifier converter is not provided"() {
+        when:
+        def measure = new PopMeasureDefinition("mid", new IdentifierObjQualifier("id"))
+        measure.withObjUriQualifiers(null)
+
+        then:
+        thrown(IllegalArgumentException)
+    }
+
+    def "should fail when uri converter cannot convert object's identifier qualifiers"() {
+        when:
+        def measure = new PopMeasureDefinition("mid", new IdentifierObjQualifier("id"))
+        measure.withObjUriQualifiers({ identifierQualifier -> Optional.empty() })
 
         then:
+        thrown(IllegalArgumentException)
+    }
+
+    def "test serializable"() {
+        given:
+        PopMeasureDefinition measureDefinition = readObjectFromResource("/$POP_MEASURE_DEFINITION_JSON", PopMeasureDefinition)
+        PopMeasureDefinition deserialized = SerializationUtils.roundtrip(measureDefinition)
+
+        expect:
         that deserialized, jsonEquals(measureDefinition)
+    }
+
+    def "should verify equals"() {
+        expect:
+        EqualsVerifier.forClass(PopMeasureDefinition)
+                .usingGetClass()
+                .verify()
+    }
 
-        where:
-        jsonMeasureDefinition << [POP_MEASURE_JSON, POP_MEASURE_WITH_OFFSET_JSON]
+    def "should return null when qualifier is requested but was not set"() {
```

Comment:
merge with the test below using `where`

Reply:
This tests two different methods: `getObjQualifier` and `getObjQualifiers`. It could be merged but I would rather leave it separate. This will lead to tests that would be easier to refactor in future when one of the methods would change.